HelpSource/Reference/Control-Structures.schelp

```diff
@@ -116,57 +125,109 @@ Do is used to iterate over a link::Classes/Collection::. Positive Integers also
+"you".do { arg item; item.postln }; // a String is a collection of characters
 
-'they'.do({ arg item; item.postln }); // a Symbol is a singular item
+'they'.do { arg item; item.postln }; // a Symbol is a singular item
 
-(8..20).do({ arg item; item.postln }); // iterates from eight to twenty
+(8..20).do { arg item; item.postln }; // iterates from eight to twenty
 
-(8,10..20).do({ arg item; item.postln }); // iterates from eight to twenty, with stepsize two
+(8,10..20).do { arg item; item.postln }; // iterates from eight to twenty, with stepsize two
 
-Routine({ var i=10; while { i > 0 } { i.yield; i = i - 5.0.rand } }).do({ arg item; item.postln });
+Routine {
+	var i = 10;
+	while { i > 0 } {
+		i.yield;
+		i = i - 5.0.rand
+	}
+}.do { arg item; item.postln };  // 'do' applies to the Routine
 ::
 
-Note:: The syntax code::(8..20).do:: uses an optimization to avoid generating an array that is used only for iteration (but which would be discarded thereafter). The return value of code:: (8..20).do({ |item| item.postln }) :: is 8, the starting value.
+Note:: The syntax code::(8..20).do:: uses an optimization to avoid generating an array that is used only for iteration (but which would be discarded thereafter). The return value of code:: (8..20).do { |item| item.postln } :: is 8, the starting value.
+
+However, if code::do:: is written as an infix binary operator, as in code::(8..20) do: { |item| item.postln }::, then it will generate the series as an array first, before calling Array:do. One side effect of this is that it is valid to code::do:: over an infinite series within a routine only if code::do:: is written as a method call. If it is written as a binary operator, you will get a "wrong type" error because the array must be finite.
+
 code::
-(8..20) do: { |item| item.postln } // is not optimized, and returns the array.
+// OK: 'do' is a method call
+r = Routine {
+	(8 .. ).do { |i|
+		i.yield;
+	};
+};
+
+r.next;
+-> 8
+-> 9 etc.
+
+// ERROR: 'do' is an operator
+r = Routine {
+	(8 .. ) do: { |i|
+		i.yield;
+	};
+};
+
+r.next;
+
+ERROR: Primitive '_SimpleNumberSeries' failed.
+Wrong type.
 ::
 ::
 
 method:: switch
 
 Object implements a switch method which allows for conditional evaluation with multiple cases. These are implemented as pairs of test objects (tested using if this == test.value) and corresponding functions to be evaluated if true.
+
 The switch statement will be inlined if the test objects are all Floats, Integers, Symbols, Chars, nil, false, true and if the functions have no variable or argument declarations. The inlined switch uses a hash lookup (which is faster than nested if statements), so it should be very fast and scale to any number of clauses.
 
+note::Hash lookup by definition implies testing emphasis::identity:: rather than equality: a code::switch:: construction that is not inlined will test code::==::, while one that is inlined will test code::===::. See the examples.::
```

Comment:
Optional: maybe this sentence could also mention that this typically means using symbols instead of strings or something like that?

Reply:
Makes sense -- done.